### PR TITLE
Implemented Muting/Unmuting of streams and topics

### LINKED ISSUE #267: Support muting/unmuting
Event needed is `muted_topics`.

- [ ] Mute topic or stream
- [ ] Unmute topic or stream

### PR DESCRIPTION
**Pull request by kunall17**
Fixes #267, opened a PR in the server zulip/zulip#3983 for implementing the routes!

### CODE REVIEW

**Comment by smarx** [comment]
*Automated message from Dropbox CLA bot*

@kunall17, it looks like you've already signed the Dropbox CLA. Thanks!

**Comment by borisyankov** [comment]
Just to update you on this. The PR is mostly good, I got few comments, but will make them later.
We are waiting on another 'long press' PR and then we'll have these PRs rebased against master.

**Comment by borisyankov** [comment]
Let's split this in few commits?
Support for messages, reducers etc. and then the UI.

**Comment by kunall17** [comment]
@borisyankov Updated the branch, this does not have the reducers. 
PR for the server also updated.

**Comment by kunall17** [comment]
The server PR just got merged in, the endpoints will be deployed in Zulip 1.6 :)

**Comment by kunall17** [comment]
@borisyankov Updated this one as well!

**Comment by borisyankov** [comment]
Great. The way you are handling the action sheet is generally correct, I just don't like the component's API and will check if the other almost similarly called component https://github.com/expo/react-native-action-sheet is a better fit (or you can check too, if you want)

**Comment by kunall17** [comment]
I'll check it out!

**Comment by kunall17** [comment]
Used expo action sheet

File: .eslintrc [hunk 1]
```diff
@@ -72,7 +72,8 @@
         "chatbubbles", "nonexisting", "primarytext", "autocomple", "stringify", "backend",
         "Menlo", "ok", "py", "todo", "Mc", "lodash", "selectable", "isequal", "lightgray", "tc",
         "zulipchat", "prepend", "pierre", "allen", "jan", "donald", "jane", "unicode", "joe",
-        "dan", "abramov", "lang", "Crashlytics", "bool", "ionicons", "truthy", "bezier", "decrement", "js", "jsonp",
+        "dan", "abramov", "lang", "Crashlytics", "bool", "ionicons", "truthy", "bezier", "decrement",
+        "jsonp", "js", "actionsheet", "ionicons", "Un", "Unmute"
```
Comment (borisyankov) [inline .eslintrc]: You will not need the 'Un' here if you rename everything from 'unMute' to 'unmute' (which is more correct)

File: src/ZulipMobile.js [hunk 2]
```diff
@@ -32,14 +33,16 @@ export default class ZulipMobile extends Component {
...
       <Provider store={store}>
-        <IntlProvider
-          key={locale}
-          locale={locale}
-          textComponent={Text}
-          messages={messages[locale]}
-        >
-          <NavigationContainer />
-        </IntlProvider>
+        <ActionSheetProvider>
```
Comment (borisyankov) [inline src/ZulipMobile.js]: Do we have to have the ActionSheetProvider here? Maybe inside the Chat.js or even MessageList?

File: src/message/MessageList.js [hunk 3]
```diff
@@ -1,17 +1,22 @@
...
-import { actionSheetButtons, executeActionSheetAction } from './messageActionSheet';
+import { constructActionButtons, executeActionSheetAction } from './messageActionSheet';
 
-export default class MessageList extends React.PureComponent {
+class MessageList extends React.PureComponent {
   autoScrollToBottom = false;
 
+  constructor(props) {
+    super();
+    this.state = { actionSheetButtons: ['', ''] };
```
Comment (borisyankov) [inline src/message/MessageList.js]: Think if you can implement this feature without using state?
I am not sure how I'll try to help with ideas.

File: src/message/messageActionSheet.js [hunk 4]
```diff
@@ -16,18 +20,60 @@ const copyToClipboard = async ({ message, auth }) => {
   Clipboard.setString(rawMessage);
 };
 
-export const actionSheetButtons = [
-  'Narrow to conversation',
-  'Reply',
-  'Copy to clipboard',
-  'Cancel'
+const unMuteTopicHelper = ({ auth, message }) => {
```
Comment (borisyankov) [inline src/message/messageActionSheet.js]: Definitely not liking 'helper' in names of functions and anything. Usually don't mean anything.

File: src/message/messageActionSheet.js [hunk 5]
```diff
@@ -16,18 +20,60 @@ const copyToClipboard = async ({ message, auth }) => {
...
+};
+
+const muteStreamHelper = ({ auth, message, subscriptions }) => {
+  const sub = subscriptions.find(x => x.name === message.display_recipient);
+  muteStream(auth, sub.stream_id);
+};
+
+const skip = () => false;
+
+const actionSheetButtons = [
```
Comment (borisyankov) [inline src/message/messageActionSheet.js]: We'll need to translate all these strings.
Ping me when you finish the other things to decide how to do it.

File: src/message/messageActionSheet.js [hunk 6]
```diff
@@ -16,18 +20,60 @@ const copyToClipboard = async ({ message, auth }) => {
...
+  { title: 'Narrow to conversation', onPress: narrowToConversation },
+  { title: 'Reply', onPress: reply },
+  { title: 'Copy to clipboard', onPress: copyToClipboard },
+  // If skip then covered in constructActionButtons
+  { title: 'Unmute Topic', onPress: unMuteTopicHelper, onlyIf: skip },
+  { title: 'Mute Topic', onPress: muteTopicHelper, onlyIf: skip },
+  { title: 'Mute Stream', onPress: muteStreamHelper, onlyIf: skip },
```
Comment (borisyankov) [inline src/message/messageActionSheet.js]: use small cases, so not 'Mute Stream' but 'Mute stream'

File: src/message/messageActionSheet.js [hunk 7]
```diff
@@ -16,18 +20,60 @@ const copyToClipboard = async ({ message, auth }) => {
...
-    () => {},
-  ][buttonIndex](props);
+export const constructActionButtons = ({ message, auth, narrow, subscriptions, mute }) => {
+  const buttons = actionSheetButtons.filter(x =>
+    !x.onlyIf || x.onlyIf({ message, auth, narrow })).map(x => x.title);
+
+  // These are dependent conditions, hence better if we manage here rather than using onlyIf
+  if (message.type === 'stream') {
+    if (isTopicMuted(message, mute)) {
+      buttons.push('Unmute Topic');
```
Comment (borisyankov) [inline src/message/messageActionSheet.js]: We will want to implement this a bit differently, so we don't use raw strings (that we need to translate).

**Comment by kunall17** [comment]
Updated :)

**Pull request merged**